src/plugins/csvstorage/csvstorage.c

```diff
@@ -0,0 +1,350 @@
+	Key *readHeaderKey = ksLookupByName(config, "/useheader", 0);
+	Key *key;
+	KeySet *header = ksNew(0, KS_END);
+	Key *dirKey;
+	Key *cur;
+	FILE *fp = NULL;
+	char *col;
+	char buf[15];
+	int nr_keys = 1;
+	
+	char delim = ';';
+	if(delimKey)
+		delim = ((char *)keyString(delimKey))[0];
+	
+	uint8_t useHeader = 0;
+	if(readHeaderKey)
+	{
+		if((((char *)keyString(readHeaderKey))[0] - '0') == 1)
+			useHeader = 1;
+	}
+	const char *fileName;
+	fileName = keyString(parentKey);
+	fp = fopen(fileName, "rb");
+	if(!fp)
+	{
+		ELEKTRA_SET_ERROR(116, parentKey, "couldn't open file");
+		return -1;
+	}
+
+	unsigned long length = 0;
+	length = getLineLength(fp);
+	if(length == 0)
+	{
+		ELEKTRA_ADD_WARNING(118, parentKey, "Empty file");
+		fclose(fp);
+		return 1;
+	}
+
+	char *lineBuffer;
+	lineBuffer = malloc(length * sizeof(char)+1);
+	if(!lineBuffer)
+	{
+		ELEKTRA_SET_ERROR(119, parentKey, "Out of memory");
+		return -1;
+	}
+	if(!fgets(lineBuffer, length, fp))
+	{
+		ELEKTRA_SET_ERROR(116, parentKey, "Cant read File");
+		return -1;
+	}
+	
+	unsigned long columns = 0;
+	columns = getColumnCount(lineBuffer, delim);
+
+	unsigned long colCounter = 0;
+	unsigned long lineCounter = 0;
+	unsigned long offset = 0;
+	if(useHeader == 1)
+	{
+		colCounter = 0;
+		offset = 0;
+		while((col = parseLine(lineBuffer, delim, offset)) != NULL)
+		{
+			offset += strlen(col)+1;
+			key = keyDup(parentKey);
+			keyAddBaseName(key, col);
+			keySetMeta(key, "csvOrder", itostr(buf, colCounter, sizeof(buf)-1));
+			ksAppendKey(header, key);
+			++colCounter;
+		}
+	}
+	else
+	{
+		colCounter = 0;
+		while(colCounter < columns)
+		{
+			key = keyDup(parentKey);
+			col = itostr(buf, colCounter, sizeof(buf)-1);
+			keyAddBaseName(key, col);
+			keySetMeta(key, "csvOrder", col);
+			ksAppendKey(header, key);
+			++colCounter;
+		}
+	}
+
+	while(!feof(fp))
+	{
+		length = getLineLength(fp);
+		if(length == 0)
+			break;
+		lineBuffer = realloc(lineBuffer, length * sizeof(char)+1);
+		fgets(lineBuffer, length, fp);
+		dirKey = keyDup(parentKey);
+		snprintf(buf, sizeof(buf)-1, "#%lu", lineCounter);
+		keyAddBaseName(dirKey, buf);
+		ksAppendKey(returned, dirKey);
+		++nr_keys;
+		offset = 0;
+		colCounter = 0;
+		while((col = parseLine(lineBuffer, delim, offset)) != NULL)
+		{
+			cur = getKeyByOrderNr(header, colCounter);
+			offset += strlen(col)+1;
+			key = keyDup(dirKey);
+			keyAddBaseName(key, keyBaseName(cur));
+			keySetString(key, col);
+			keySetMeta(key, "csvOrder", itostr(buf, colCounter, sizeof(buf)-1));
+			ksAppendKey(returned, key);
+			keyDel(key);
+			++nr_keys;
+			++colCounter;
+		}
+		if(colCounter != columns)
+		{
+			ELEKTRA_ADD_WARNING(118, parentKey, "illegal number of columns");
+		}
+		++lineCounter;
+		keyDel(dirKey);
+	}
+
+	key	= keyDup(parentKey);
+	snprintf(buf, sizeof(buf)-1, "#%lu", lineCounter);
+	keySetString(key, buf);
+	ksAppendKey(returned, key);
+	
+	fclose(fp);
+	free(lineBuffer);
+	ksDel(header);
+	return nr_keys;
+}
```

Comment:
I'm still searching for a solution that doesn't basically require me to pass all variables to a function and turn 5 lines of code into 20.